### ISSUE TITLE
Sum style composition via types or via explicit operators?

### ISSUE THREAD

**Issue by harendra-kumar** [composewell/streamly#18]
Currently we use types (StreamT, ParallelT etc.) to determine how streams compose in the product composition. However, for sum style composition we use explicit operators i.e. `<>`, `<=>`, `<|` and `<|>` to combine streams. They combine streams in operator specific manner irrespective of the type (i.e. StreamT, ParallelT ect.).  The MonadPlus instance for all types is the same, it combines the streams in parallel using `<|>`. 

This issue is about considering to change the MonadPlus instance and make it specific to the type so that we have a type specific way of combining the streams in a sum style. It will allow us to use msum on foldables to turn them into streams based on the type. We can also choose an operator for `mplus` e.g. `<.>` or `<+>` consistent with other operators but combining in a type specific way. The type specific operator will allow us to just change the type annotation to change the behavior. Whereas the type independent operators fix the behavior. Both of these may be necessary for different use cases.

Since this is a breaking change, better to do this earlier than later if we decide to do it.

**Comment by harendra-kumar** [comment]
Another choice for the type specific sum operator could be `<:>`.

**Comment by harendra-kumar** [comment]
see #20, we would like to use the Alternative operator `<|>` for choice or race like behavior. And, we should keep the MonadPlus as well compatible with the same.

I will summarize my current thinking about operators and typeclass instances:

* <|> : This is standard Alternative type class with a standard alternative behavior for serial types (StreamT and InterleavedT) and race like behavior for parallel types (i.e. AsyncT and ParallelT).
* <> : Use the semigroup instance to add streams in different ways. Currently this is type independent and adds serially. However we can make it type specific. For StreamT it will be the same as now. For InterleavedT it will be same as `<=>` or `interleave`, for AsyncT it will be `<|`, for ParallelT it will be what is `<|>` now i.e. all actions running in parallel. All these form a semigroup as long as the equality is multiset equality rather than ordered-list equality.

We need to make the following changes:
* We need to change the `<>` and make it type specific. This should be done early as this is a breaking change.
* We need to implement the new `<|>` for choice behavior. This can be done as a separate change later as this is not really urgent and is a new functionality. We may just remove the Alternative instance for now.

With this change we can use something like this to choose the correct instance:

```
serially $ return 1 <> return 2 <> return 3
asyncly $ return 1 <> return 2 <> return 3
parallely $ return 1 <> return 2 <> return 3
```

The behavior of the stream is governed by the combinator using the same type-independent sum operator. This is a breaking change as we are changing the meaning of `<>` and `<|>`. We should deprecate the `<|` and `<=>` operators in favor of using the combinator and `<>`.

I am not in favor of keeping type specific operators as we may require too many if we have to represent all types of Alternatives, Semigroups, Applicatives, Zips and many other types to come. However, for the record here is a set of operators that I thought about so that they have a convention and are consistently named:

```
Sum Style (`+` indicates sum, `<>` indicate serial, `|` indicates parallel):

Serial           +>
Interleaved     <+>
Ahead           +|
Async           +||
Parallel        |+|

Alternatives (`.` indicates alternative)

Serial           .>
Interleaved     <.>
Ahead           .|
Async           .||
Parallel        |.|

Zips

Serial   ><
Async    >|<

Note that there are other versions of zip possible, one corresponding to each applicative style.
```

**Comment by harendra-kumar** [comment]
Note that, the race combinator `<|>` can be implemented by just doing everything in parallel and use take 1 which should cancel the pending tasks.

**Comment by harendra-kumar** [comment]
I am in favor of removing the `Alternative` instance altogether and use the `take 1` mechanism to implement the racing behavior, that is much cleaner.

**Comment by harendra-kumar** [comment]
The race behavior does not fit the Alternative instance of a stream, it is incorrect to implement that using Alternative. However, a perfectly valid and potentially useful alternative instance is possible for streams, the Alternative instance must be such that an empty stream represents `empty` or failure and a non-empty stream represents success. The alternative should compose such that it returns the first non-empty stream.

**Comment by harendra-kumar** [comment]
This has been addressed via #60. We now compose in sum style using semigroup instance of each typeclass. We also have polymorphic functions to compose in a particular way without having to change the type of the stream.
